### PR TITLE
Use Docker events API instead of polling Docker daemon.

### LINKED ISSUE #84: Consider using docker events API instead of polling in docker process mapper.

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #84

### CODE REVIEW

File: probe/docker_process_mapper.go [hunk 1]
```diff
@@ -50,63 +70,145 @@ var (
...
+		case <-otherUpdates:
+			if err := m.updatePIDTree(); err != nil {
+				log.Printf("docker mapper: %s", err)
+				return
+			}
+
+			if err := m.updateImages(client); err != nil {
+				log.Printf("docker mapper: %s", err)
+				return
+			}
```
Comment (peterbourgon) [inline probe/docker_process_mapper.go]: Do you really want to return in these error cases? Could you continue instead?

File: probe/docker_process_mapper.go [hunk 2]
```diff
@@ -50,63 +70,145 @@ var (
...
-	pmap := map[int]*docker.Container{}
-	for _, container := range containers {
-		info, err := client.InspectContainer(container.ID)
+	if err := m.updateImages(client); err != nil {
+		log.Printf("docker mapper: %s", err)
+		return
+	}
+
+	otherUpdates := time.Tick(m.interval)
+	for true {
```
Comment (peterbourgon) [inline probe/docker_process_mapper.go]: `true` is unnecessary. Also, should this thing have a Stop mechanism?

File: probe/docker_process_mapper.go [hunk 3]
```diff
@@ -50,63 +70,145 @@ var (
...
+		m.images[image.ID] = image
+	}
+	m.Unlock()
+
+	return nil
+}
+
+func (m *dockerMapper) handleEvent(event *docker.APIEvents, client dockerClient) {
+	switch event.Status {
+	case "stop":
```
Comment (peterbourgon) [inline probe/docker_process_mapper.go]: Can we use a constant here instead of a string literal?

File: probe/docker_process_mapper.go [hunk 4]
```diff
@@ -50,63 +70,145 @@ var (
...
+		containerID := event.ID
+		m.Lock()
+		container, ok := m.containers[containerID]
+		if ok {
+			delete(m.containers, containerID)
+			delete(m.containersByPID, container.State.Pid)
+		} else {
+			log.Printf("docker mapper: container %s not found", containerID)
+		}
```
Comment (peterbourgon) [inline probe/docker_process_mapper.go]:
```go
if container, ok := m.containers[event.ID]; ok {
    delete(m.containers, event.ID)
    delete(m.containersByPID, container.State.Pid)
} else {
    log.Printf("...")
}
```

File: probe/docker_process_mapper_test.go [hunk 5]
```diff
@@ -59,12 +67,14 @@ func TestDockerProcessMapper(t *testing.T) {
...
 	}
 
-	dockerMapper := newDockerMapper("/proc", 10*time.Second)
+	dockerMapper, _ := newDockerMapper("/proc", 10*time.Second)
 	dockerIDMapper := dockerMapper.idMapper()
 	dockerNameMapper := dockerMapper.nameMapper()
 	dockerImageIDMapper := dockerMapper.imageIDMapper()
 	dockerImageNameMapper := dockerMapper.imageNameMapper()
 
+	<-time.After(time.Millisecond)
```
Comment (peterbourgon) [inline probe/docker_process_mapper_test.go]:
```go
time.Sleep(time.Millisecond)
```

Or consider using

```go
runtime.Gosched()
```

**Comment by peterbourgon** [comment]
>  coverage/coveralls — Coverage decreased (-1.79%) to 62.24%

File: probe/docker_process_mapper.go (hunk 1 above)
Comment (tomwilkie) [inline probe/docker_process_mapper.go]: Yeah its probably not important, the whole thing is in a big while loop anyway.

File: probe/docker_process_mapper.go (hunk 2 above)
Comment (tomwilkie) [inline probe/docker_process_mapper.go]: Yes we discussed this a while back, but I forgot to file a ticket.  This PR doesn't make the situation worse, so do you mind if we defer a Stop mechanism to another day?

Comment (tomwilkie) [inline probe/docker_process_mapper.go]: #166

File: probe/docker_process_mapper.go (hunk 4 above)
Comment (tomwilkie) [inline probe/docker_process_mapper.go]: Done.

**Comment by tomwilkie** [comment]
Thanks for feedback!  PTAL.  I'm not to fussed about small drops in coverage like that; majority is error handling.

File: probe/docker_process_mapper.go (hunk 2 above)
Comment (peterbourgon) [inline probe/docker_process_mapper.go]: Sounds fine to me.

**Comment by peterbourgon** [comment]
:+1:

**Pull request merged**